### PR TITLE
Implementation of the `const_static_lifetime` lint.

### PR DESCRIPTION
**Pull request by gbip**
Hello,
After a few month, I decided to come back on some unfinished work (previous pr : #1829, issue : #1762).
Currently, the lint has the correct behaviour, but I am struggling setting up the test.

`./tests/ui/const_static_lifetime.rs` contains some tests, but I can't figure out how to generate a correct .stderr output: `./target/debug/test_build_base/const_static_lifetime.err` contains hardcoded strings for file location (`tests/ui/const_static_lifetime.rs`), whereas other tests contains `$DIR/<real_filename>`.
Moreover, I have broken two test : `./tests/ui/regex.rs` that I fixed (it contained a constant with an explicit static lifetime) and `./tests/ui/for_loop.rs`. I can't figure what broke on this done...

Anyway, if some of you want to give some more test cases, I will be glad to add them in the test file.

Thanks for your time !

PS: Please be tolerant, english is not my native language, I am trying my best to avoid mistakes.

### CODE REVIEW

**Comment by oli-obk** [comment]
You need to run `./tests/ui/update_all_references.sh` after `cargo test` fails. Then you can commit the changes.

**Comment by gbip** [comment]
Hey, I just fixed the test thing !

File: clippy_lints/src/approx_const.rs [hunk 1]
```diff
@@ -33,24 +33,22 @@ declare_lint! {
 }
 
 // Tuples are of the form (constant, name, min_digits)
-const KNOWN_CONSTS: &'static [(f64, &'static str, usize)] = &[
```
Comment (oli-obk) [inline clippy_lints/src/approx_const.rs]: Please don't make unrelated formatting changes

File: clippy_lints/src/approx_const.rs [hunk 2]
```diff
@@ -83,17 +81,12 @@ fn check_known_consts(cx: &LateContext, e: &Expr, s: &symbol::Symbol, module: &s
...
         for &(constant, name, min_digits) in KNOWN_CONSTS {
             if is_approx_const(constant, &s, min_digits) {
-                span_lint(
-                    cx,
-                    APPROX_CONSTANT,
-                    e.span,
-                    &format!(
-                        "approximate value of `{}::consts::{}` found. \
-                         Consider using it directly",
-                        module,
```
Comment (oli-obk) [inline clippy_lints/src/approx_const.rs]: more unrelated formatting changes

File: clippy_lints/src/const_static_lifetime.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                }
+            },
+            // This is what we are looking for !
+            TyKind::Rptr(ref optional_lifetime, ref borrow_type) => {
+                // Match the 'static lifetime
+                if let Some(lifetime) = *optional_lifetime {
+                    if let TyKind::Path(_, _) = borrow_type.ty.node {
+                        // Verify that the path is a str
+                        if lifetime.ident.name == "'static" {
+                            span_help_and_lint(cx,
```
Comment (oli-obk) [inline clippy_lints/src/const_static_lifetime.rs]: Can you use `span_help_and_then` and add a suggestion to remove the lifetime? (Search for `span_suggestion` for examples)

**Comment by gbip** [comment]
I think everything you have requested has been done, can you check ?

File: tests/ui/regex.rs [hunk 4]
```diff
@@ -27,23 +27,11 @@ fn syntax_error() {
...
-
-    let set_error = RegexSet::new(&[
-        OPENING_PAREN,
-        r"[a-z]+\.(com|org|net)",
-    ]);
-    let bset_error = BRegexSet::new(&[
-        OPENING_PAREN,
-        r"[a-z]+\.(com|org|net)",
-    ]);
+    let set = RegexSet::new(&[r"[a-z]+@[a-z]+\.(com|org|net)", r"[a-z]+\.(com|org|net)"]);
```
Comment (oli-obk) [inline tests/ui/regex.rs]: Some leftover formatting changes

File: clippy_lints/src/const_static_lifetime.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            TyKind::Rptr(ref optional_lifetime, ref borrow_type) => {
+                // Match the 'static lifetime
+                if let Some(lifetime) = *optional_lifetime {
+                    if let TyKind::Path(_, _) = borrow_type.ty.node {
+                        // Verify that the path is a str
+                        if lifetime.ident.name == "'static" {
+                            let mut sug: String = String::new();
+                            // This should be ok, since we know that the current type is a borrow.
+                            sug.push('&');
+                            sug += &snippet(cx, borrow_type.ty.span, "<expr>").into_owned();
```
Comment (oli-obk) [inline clippy_lints/src/const_static_lifetime.rs]: The suggestion should be empty. It is the code that is written over `'static`, not over `&'static T`.

File: tests/ui/for_loop.stdout [hunk 6]
```diff
@@ -0,0 +1 @@
+array
```
Comment (oli-obk) [inline tests/ui/for_loop.stdout]: Accidentally commited debug output?

File: clippy_lints/src/const_static_lifetime.rs (hunk 5 above)
Comment (gbip) [reply clippy_lints/src/const_static_lifetime.rs]: Oh, thanks I couldn't figured it out.
Is this for the application that automatically applies suggestions ?

Comment (oli-obk) [reply clippy_lints/src/const_static_lifetime.rs]: Yea. In the future rls will probably make use of it, too. So you'll be able to fix these things with a single click on the offending error message.

Comment (gbip) [reply clippy_lints/src/const_static_lifetime.rs]: I am currently applying your requests, but `cargo test` fails because :
```
error: Unnecessary `!=` operation
   --> clippy_lints/src/loops.rs:754:21
    |
754 | /                     if offset_left.var_name != offset_right.var_name {
755 | |                         Some((offset_left, offset_right))
756 | |                     } else {
757 | |                         None
758 | |                     }
    | |_____________________^
    |
    = note: `-D if-not-else` implied by `-D clippy-pedantic`
    = help: change to `==` and swap the blocks of the if/else
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.166/index.html#if_not_else

error: aborting due to previous error
```
Is this related to the latest nigthly or some error in my changes ?

Comment (oli-obk) [reply clippy_lints/src/const_static_lifetime.rs]: curious. I wonder why travis didn't catch that on the other PRs... I'll push a quick fix that you can rebase over

Comment (oli-obk) [reply clippy_lints/src/const_static_lifetime.rs]: You can rebase over the master branch, I fixed it

**Comment by oli-obk** [comment]
Something really weird happened. you have all of our commits in this PR now.

Try `git rebase -i origin/master` and remove all commits that aren't yours

**Comment by gbip** [comment]
Thanks, I was looking into it !

**Comment by oli-obk** [comment]
Thanks! This lint is so much in the spirit of clippy. Annoying as hell 😆 I will keep doing it wrong until clippy beats `'static` out of my muscle memory

**Comment by gbip** [comment]
I really enjoyed implementing this first lint, thanks for your help !
